Use some if constexpr to reduce the code

aprokop opened the pull request:
@dalg24 wanted to run this past you before creating a public PR.

dalg24 commented on src/details/ArborX_DetailsMortonCode.hpp:
```diff
@@ -192,47 +192,35 @@ KOKKOS_INLINE_FUNCTION unsigned int morton32(Point const &p)
+  constexpr int DIM = GeometryTraits::dimension_v<Point>;
+  constexpr unsigned N = (1u << (63 / DIM));
+
+  if constexpr (DIM > 2)
```
Original code was doing `DIM != 2`, please confirm you meant to change this and make sure the upcoming PR description mentions it if the answer is yes.

aprokop replied:
I added static asserts to make sure `DIM >= 2`. We don't support (and don't think we should) `DIM = 1` case.

dalg24 commented on src/details/ArborX_DetailsMortonCode.hpp:
```diff
@@ -192,47 +192,35 @@ KOKKOS_INLINE_FUNCTION unsigned int morton32(Point const &p)
+    // Special case because it needs double. Float is not sufficient to
+    // represent large integers, which would result in some missing bins.
+    auto xd = min(max((double)p[0] * N, 0.), (double)N - 1);
+    auto yd = min(max((double)p[1] * N, 0.), (double)N - 1);
```
Would def not be right for `DIM==1`

dalg24 commented on src/geometry/ArborX_HyperBox.hpp:
```diff
@@ -68,32 +68,28 @@ struct Box
+      }
+      else
+      {
+        assert(false);
```
```suggestion
        static_assert(std::is_void_v<Geometry>, "current implementation only supporting point or box");
```

dalg24 commented on src/details/ArborX_DetailsUtils.hpp:
```diff
@@ -74,18 +74,44 @@ create_layout_right_mirror_view(
-  return src;
+inline auto
+create_layout_right_mirror_view(ExecutionSpace const & /*execution_space*/,
+                                View const &src)
```
This is fine but it reminds me how this function could use some refactor/cleanup

dalg24 commented on src/details/ArborX_SpaceFillingCurves.hpp:
```diff
@@ -27,43 +27,37 @@ namespace Experimental
                                   Geometry const &geometry) const
   {
-    auto p = Details::returnCentroid(geometry);
+    auto p = [&geometry]() {
+      if constexpr (GeometryTraits::is_point<Geometry>{})
+        return geometry;
+      else
+        return Details::returnCentroid(geometry);
+    }();
```
I really dislike this.  Do we really need the code specialization?

aprokop committed:
Switch host in HIP testing from Serial to OpenMP

aprokop committed:
Merge pull request #794 from aprokop/hip_build

aprokop committed:
Switch google benchmark rates to use 1.5 rate feature

aprokop committed:
Update gitlab-ci.yml to install benchmark 1.6.0

aprokop committed:
Require Google benchmark 1.5.4

- This is the first version introduced proper CMake version support

aprokop committed:
Upgrade SYCL container's benchmark to 1.5.4

aprokop committed:
[ci skip] Update nightly containers to Ubuntu 22.04

dalg24 committed:
Drop pre Kokkos 3.6 PTHREAD workaround

aprokop committed:
Merge pull request #800 from dalg24/drop_pthread_workaround

dalg24 committed:
Fixup boost::program_options::bool_switch do not take a default value

dalg24 committed:
Do not bother with downloading GTest

aprokop committed:
Merge pull request #803 from dalg24/gilab_gtest

Fix https://github.com/arborx/ArborX/pull/799#discussion_r1058338594

dalg24 committed:
Merge pull request #804 from dalg24/bool_switch_default_value

Fix undefined behavior when parsing boolean options in DBSCAN benchmark

aprokop committed:
Add testing for MST with mutual reachability metric (#802)

* Add testing for MST with mutual reachability metric

Testing mrd based on the edges output is hard because many edges may
have the same weight. Instead, we just test the total weight of the MST,
and compare to the reference weight computed externally.

* Switch back to using relative tolerance to get better error message

* Add missing <map> header

dalg24 committed:
Merge pull request #799 from aprokop/require_benchmark_1.5

Require benchmark 1.5.4, update nightly containers to Ubuntu 22.04

dalg24 committed:
Annotate fence before return statement in KokkosExt::lastELement

dalg24 committed:
Annotate global fence in Details::dendrogramUnionFindHost

dalg24 committed:
Annotate fences in DistributedTree

aprokop committed:
Merge pull request #805 from dalg24/annotate_fences

aprokop committed:
Use constexpr to cut down on the number of unncessary templated functions
